lib/winapi/badptr.c

```diff
@@ -0,0 +1,9 @@
+#include <windows.h>
+
+BOOL IsBadWritePtr(LPVOID lp, UINT_PTR ucb)
+{
+  //FIXME: Run NtQueryVirtualMemory for the pages?
```

Comment:
The implementation on Windows registers an SEH handler (checked in Ghidra, but the MSDN description already hints it) and then just write probes each page. I didn't check the actual handler, but it's safe to assume it just catches the exception and returns `TRUE`.

As for safety on the Xbox: Write probing can't really be safe here (the pointer could point to MMIO or something). I don't think `NtQueryVirtualMemory` can solve the problem though, since the pages that make up the memory to be checked might not belong to the same allocation. I think the best we can do is replicate the write probing once we have SEH support, and add a warning about this function being unsafe.

The stub is fine for now, but I recommend creating an issue about this after merge.

Reply:
I'll add a note about the SEH handler implementation on Windows.

---

My idea would have been to just keep looping until all all `(ucb + 0xFFF) / 0x1000` pages starting at `lp & ~0xFFF` have been checked. If any of them doesn't have write permission, report that.
When implementing this algorithm with `NtQueryVirtualMemory` I'd skip `RegionSize` bytes at a time (instead of walking each page, I'd just walk allocations).

It would at least do what MSDN says:

> If the calling process has write access to all bytes in the specified memory range, the return value is zero.

Hence, in addition to the SEH comment, I'd like to keep the existing comment. I'll improve it to say something like "//FIXME: Walk all allocations / pages in the region with NtQueryVirtualMemory".

It was already my plan to create an issue to document this being a stub.

Reply:
> My idea would have been to just keep looping until all all `(ucb + 0xFFF) / 0x1000` pages starting at `lp & ~0xFFF` have been checked. If any of them doesn't have write permission, report that.
> When implementing this algorithm with `NtQueryVirtualMemory` I'd skip `RegionSize` bytes at a time (instead of walking each page, I'd just walk allocations).

What I meant is that I'm not sure `NtQueryVirtualMemory()` can do that. MSDN mentions `state, protection, and type of a region of pages`, which, to me, sounds like it's only meant to be called for addresses that mark the begin of an allocation done by `NtAllocateVirtualMemory()`. This is only a suspicion though, based on the MSDN wording and the fact that it only takes a base address and no size / page count parameter. I did not check the actual behavior.